File: frontend/src/lib/attack.ts
// ATT&CK tactic mapping — mirrors crates/sigil-correlate/src/attack.rs so the
// UI can group techniques into tactics without an extra API call.

export const TACTICS = [
  'initial-access',
  'execution',
  'persistence',
  'privilege-escalation',
  'defense-evasion',
  'credential-access',
  'discovery',
  'lateral-movement',
  'collection',
  'command-and-control',
  'exfiltration',
  'impact',
] as const;

export function tacticFor(technique?: string | null): string {
  if (!technique) return 'unknown';
  const t = technique.toUpperCase();
  const has = (...p: string[]) => p.some((x) => t.startsWith(x));
  if (has('T1110', 'T1003', 'T1552')) return 'credential-access';
  if (has('T1548', 'T1068', 'T1078')) return 'privilege-escalation';
  if (has('T1059', 'T1203')) return 'execution';
  if (has('T1071', 'T1572', 'T1105')) return 'command-and-control';
  if (has('T1041', 'T1048')) return 'exfiltration';
  if (has('T1021', 'T1210')) return 'lateral-movement';
  return 'unknown';
}

File: crates/sigil-correlate/src/attack.rs
//! ATT&CK tactic mapping (DESIGN §9.6): map a technique id (or, failing that,
//! the event's OCSF class) to a tactic, so a chain of events becomes a
//! tactic→technique kill-chain. This is the honest, campaign-level attribution
//! of §9.7 — not named-actor attribution.

use sigil_core::OcsfClass;

/// Tactic for a technique id (e.g. `T1110.001` → `credential-access`), falling
/// back to a class-based default when the technique is unknown.
pub fn tactic_for(technique: Option<&str>, class: &OcsfClass) -> &'static str {
    if let Some(t) = technique {
        let t = t.to_ascii_uppercase();
        let by_technique =
            if t.starts_with("T1110") || t.starts_with("T1003") || t.starts_with("T1552") {
                Some("credential-access")
            } else if t.starts_with("T1548") || t.starts_with("T1068") || t.starts_with("T1078") {
                Some("privilege-escalation")
            } else if t.starts_with("T1059") || t.starts_with("T1203") {
                Some("execution")
            } else if t.starts_with("T1071") || t.starts_with("T1572") || t.starts_with("T1105") {
                Some("command-and-control")
            } else if t.starts_with("T1041") || t.starts_with("T1048") {
                Some("exfiltration")
            } else if t.starts_with("T1021") || t.starts_with("T1210") {
                Some("lateral-movement")
            } else {
                None
            };
        if let Some(tac) = by_technique {
            return tac;
        }
    }
    tactic_for_class(class)
}

/// Coarse tactic guess from the OCSF class alone.
pub fn tactic_for_class(class: &OcsfClass) -> &'static str {
    match class {
        OcsfClass::Authentication => "credential-access",
        OcsfClass::ProcessActivity => "execution",
        OcsfClass::NetworkActivity => "command-and-control",
        OcsfClass::FileSystemActivity => "collection",
        OcsfClass::HttpActivity => "command-and-control",
        OcsfClass::ApiActivity => "execution",
        OcsfClass::DnsActivity => "command-and-control",
        OcsfClass::ModuleActivity => "defense-evasion",
        OcsfClass::ScheduledJobActivity => "persistence",
        OcsfClass::RegistryKeyActivity => "persistence",
        OcsfClass::Other(_) => "unknown",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn technique_maps_to_tactic() {
        assert_eq!(
            tactic_for(Some("T1110.001"), &OcsfClass::Authentication),
            "credential-access"
        );
        assert_eq!(
            tactic_for(Some("T1548.003"), &OcsfClass::ProcessActivity),
            "privilege-escalation"
        );
    }

    #[test]
    fn falls_back_to_class() {
        assert_eq!(
            tactic_for(None, &OcsfClass::NetworkActivity),
            "command-and-control"
        );
        assert_eq!(
            tactic_for(Some("T9999"), &OcsfClass::ProcessActivity),
            "execution"
        );
    }
}
